### PR TITLE
Catch yapf exceptions

### LINKED ISSUE #901: yapf support is fragile in the face of a file with a syntax error
When using `C-c C-r f` to format a file, with yapf installed (0.9.0 w/ Python3), and a simple python file containing a syntax error, yapf throws an exception:

```
~/tmp/test : [master x] yapf --style pep8 --no-local-style functional_tests.py
Traceback (most recent call last):
  File "/usr/local/bin/yapf", line 11, in <module>
    sys.exit(run_main())
  File "/usr/local/lib/python2.7/site-packages/yapf/__init__.py", line 248, in run_main
    sys.exit(main(sys.argv))
  File "/usr/local/lib/python2.7/site-packages/yapf/__init__.py", line 166, in main
    verify=args.verify)
  File "/usr/local/lib/python2.7/site-packages/yapf/__init__.py", line 210, in FormatFiles
    logger=logging.warning)
  File "/usr/local/lib/python2.7/site-packages/yapf/yapflib/yapf_api.py", line 89, in FormatFile
    verify=verify)
  File "/usr/local/lib/python2.7/site-packages/yapf/yapflib/yapf_api.py", line 122, in FormatCode
    tree = pytree_utils.ParseCodeToTree(unformatted_source)
  File "/usr/local/lib/python2.7/site-packages/yapf/yapflib/pytree_utils.py", line 113, in ParseCodeToTree
    raise e
  File "functional_tests.py", line 20
    msg='Browser title was {0}'.format(browser.title))
      ^
SyntaxError: invalid syntax
```

This then causes Elpy's yapf support to halt unpleasantly, with a "wrong-type-argument char-or-string-p nil" error that's not really explanatory:
```
Debugger entered--Lisp error: (wrong-type-argument char-or-string-p nil)
  insert(nil)
  (save-excursion (goto-char end) (insert rep) (delete-region beg end))
  (if (string-equal (buffer-substring beg end) rep) nil (save-excursion (goto-char end) (insert rep) (delete-region beg end)))
  elpy-buffer--replace-region(1 1285 nil)
  (let ((new-block (elpy-rpc method (vector (elpy-rpc--buffer-contents)))) (beg (point-min)) (end (point-max))) (elpy-buffer--replace-region beg end new-block))
  (if (use-region-p) (let ((new-block (elpy-rpc method (list (elpy-rpc--region-contents)))) (beg (region-beginning)) (end (region-end))) (elpy-buffer--replace-region beg end new-block)) (let ((new-block (elpy-rpc method (vector (elpy-rpc--buffer-contents)))) (beg (point-min)) (end (point-max))) (elpy-buffer--replace-region beg end new-block)))
  (let ((line (line-number-at-pos)) (col (current-column))) (if (use-region-p) (let ((new-block (elpy-rpc method (list (elpy-rpc--region-contents)))) (beg (region-beginning)) (end (region-end))) (elpy-buffer--replace-region beg end new-block)) (let ((new-block (elpy-rpc method (vector (elpy-rpc--buffer-contents)))) (beg (point-min)) (end (point-max))) (elpy-buffer--replace-region beg end new-block))) (forward-line (1- line)) (forward-char col))
  elpy--fix-code-with-formatter("fix_code_with_yapf")
  elpy-yapf-fix-code()
  funcall-interactively(elpy-yapf-fix-code)
  call-interactively(elpy-yapf-fix-code record nil)
  command-execute(elpy-yapf-fix-code record)
  execute-extended-command(nil "elpy-yapf-fix-code" "elpy-yapf")
  funcall-interactively(execute-extended-command nil "elpy-yapf-fix-code" "elpy-yapf")
  call-interactively(execute-extended-command nil nil)
  command-execute(execute-extended-command)
```

It'd be useful if the yapf exception was gracefully handled so that the code skips over attempting to use it and continues on with autopep8 (if present). The autopep8 formatter was able to consume the same file and produce some output.

Took me a while to realize that it was the underlying yapf script that was throwing the exception and this was causing difficulties for Elpy.

### PR DESCRIPTION
**Pull request by ChillarAnand**
closes #901

### CODE REVIEW

**Comment by jorgenschaefer** [comment]
Awesome, thank you. Can you write a test for this? I.e. mock the yapf call, make it throw an error, and ensure `fix_code` raises an xmlrpclib.Fault instead?

**Comment by ChillarAnand** [comment]
Added a test case :-)

**Comment by jorgenschaefer** [comment]
Oh, nice, thank you! I did meant a Python test case – I try to keep the Python code as well-tested as possible, as it's easier to test than Elisp …

File: elpy/tests/test_yapf.py [hunk 1]
```diff
@@ -15,6 +15,10 @@ def setUp(self):
         if not yapfutil.YAPF_NOT_SUPPORTED:
             raise unittest.SkipTest
 
+    def test_fix_code_should_throw_error_for_invalid_code(self):
+        src = 'x = '
+        self.assertRaises(SyntaxError, yapfutil.fix_code, src)
+
```
Comment (ChillarAnand) [inline elpy/tests/test_yapf.py]: Coverage still decreases. This should run `except` case?

Comment (jorgenschaefer) [inline elpy/tests/test_yapf.py]: Yes … but shouldn't that code now raise `Fault`?

**Comment by ChillarAnand** [comment]
Done.

On Friday 10 June 2016, Coveralls <notifications@github.com> wrote:

> [image: Coverage Status] <https://coveralls.io/builds/6542054>
>
> Coverage increased (+0.7%) to 90.429% when pulling *23fa66f
> <https://github.com/jorgenschaefer/elpy/commit/23fa66f3676a3a36fe0ee635039add3b1a8eb33b>
> on ChillarAnand:yapf* into *45ad707
> <https://github.com/jorgenschaefer/elpy/commit/45ad70762abfecfcf1497440d4fa0715d326655c>
> on jorgenschaefer:master*.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/jorgenschaefer/elpy/pull/902#issuecomment-225115187>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AEQctKL2VkuNUg-zLy3f4Ts5JmgrxTPnks5qKRZLgaJpZM4IsyaG>
> .
>


-- 
Regards,
Chillar Anand
www.avilpage.com

**Comment by jorgenschaefer** [comment]
Excellent work, thank you!

**Pull request merged**